include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,383 @@
+};
+
+template<class F, typename Spectral>
+struct CookTorranceParams<F,Spectral,true>
+{
+    using scalar_type = typename F::scalar_type;
+    using spectral_type = Spectral;
+
+    scalar_type ax;
+    scalar_type ay;
+    F fresnel;
+};
+}
+
+// N (NDF), F (fresnel)
+template<class Config, class N, class F NBL_PRIMARY_REQUIRES(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
+struct SCookTorrance
+{
+    MICROFACET_BXDF_CONFIG_TYPE_ALIASES(Config);
+
+    using this_t = SCookTorrance<Config, N, F>;
+    using quant_type = typename N::quant_type;
+    using ndf_type = N;
+    using fresnel_type = F;
+
+    NBL_CONSTEXPR_STATIC_INLINE bool IsAnisotropic = ndf_type::IsAnisotropic;
+    NBL_CONSTEXPR_STATIC_INLINE bool IsBSDF = ndf_type::IsBSDF;
+
+    using creation_params_type = impl::CookTorranceParams<fresnel_type, spectral_type, IsAnisotropic>;
+
+    template<typename C=bool_constant<!IsAnisotropic> >
+    static enable_if_t<C::value && !IsAnisotropic, this_t> create(NBL_CONST_REF_ARG(creation_params_type) params)
+    {
+        this_t retval;
+        retval.ndf = ndf_type::create(params.A);
+        retval.fresnel = params.fresnel;
+        return retval;
+    }
+    template<typename C=bool_constant<IsAnisotropic> >
+    static enable_if_t<C::value && IsAnisotropic, this_t> create(NBL_CONST_REF_ARG(creation_params_type) params)
+    {
+        this_t retval;
+        retval.ndf = ndf_type::create(params.ax, params.ay);
+        retval.fresnel = params.fresnel;
+        return retval;
+    }
+
+    template<class Interaction, class MicrofacetCache>
+    spectral_type __eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, interaction.getNdotV());
+        const bool notTIR = impl::check_TIR_helper<fresnel_type, IsBSDF>::template __call<MicrofacetCache>(_f, cache);
+        if ((IsBSDF && notTIR) || (!IsBSDF && _sample.getNdotL() > numeric_limits<scalar_type>::min && interaction.getNdotV() > numeric_limits<scalar_type>::min))
+        {
+            using quant_query_type = typename ndf_type::quant_query_type;
+            using g2g1_query_type = typename ndf_type::g2g1_query_type;
+
+            scalar_type dummy;
+            quant_query_type qq = ndf.template createQuantQuery<MicrofacetCache>(cache, dummy);
+
+            quant_type D = ndf.template D<sample_type, Interaction, MicrofacetCache>(qq, _sample, interaction, cache);
+            scalar_type DG = D.projectedLightMeasure;
+            if (D.microfacetMeasure < bit_cast<scalar_type>(numeric_limits<scalar_type>::infinity))
+            {
+                g2g1_query_type gq = ndf.template createG2G1Query<sample_type, Interaction>(_sample, interaction);
+                DG *= ndf.template correlated<sample_type, Interaction>(gq, _sample, interaction);
+            }
```

Comment:
hmm actually since Eval is not Quotient, it should return 0 whenever PDF would have been INF, and therefore whenever Eval is inf, you can just return 0